### PR TITLE
[tests] Add support for running EndToEnd tests on helix with the `aspire` workload

### PR DESCRIPTION
**Pull request by radical**
# Problem:

1. We want to run some EndToEnd tests on CI, which can `dotnet run` an aspire project,
and allow individual tests to interact with the services.
This requires:

	- Ability to build, and run an aspire project - IOW, a sdk with the `aspire` workload installed.
	- `docker`

2. Also, allow using `TestProject.*` in `tests/testproject`, in two modes:
    - `in-tree` test run which directly reference aspire projects, and repo targets
    - `out-of-tree` test runs which uses a SDK+workload based on local build output

# Solution:

### SDK+workload

- SDK with version from `global.json` is installed into `artifacts/bin/dotnet-{none,latest}`
    - The `aspire` workload is installed into the sdk using the nuget packages in `artifacts/packages` with a version like `8.0.0-{dev,ci}`
- this also allows running tests that do `dotnet new aspire-starter; dotnet run` same as a user
- this utilizes an internal nuget from `dotnet/runtime`

### TestProject

- This can switch between the two test run modes using the msbuild property `$(TestsRunningOutsideRepo)`
    - when running `in-tree` the test project directly references hosting targets, and aspire projects via `ProjectReference`
    - when running `out-of-tree` the `ProjectReferences` and imports are replaced with `PackageReferences` to the Aspire nugets
- Default is to run `in-tree`

### Helix

- The tests are built on the build machine
- The testproject, and the sdk+workload is sent to helix
  - where the tests run using `dotnet` from the sdk+workload path
- Since `docker` is needed to helix, this is enabled only for `Linux` in this PR. Blocked on https://github.com/dotnet/dnceng/issues/2067 for windows support.

### Steps to prepare for `out-of-tree` runs

1. `.\build.cmd -pack` - to build all the nugets
2. `dotnet build tests/Aspire.EndToEnd.Tests/Aspire.EndToEnd.csproj /t:InstallWorkloadUsingArtifacts /p:Configuration=<config>`
    - this will install the sdk, and the `aspire` workload using the nugets from `artifacts/packages`

- at this point the sdk is usable outside the repo too. Note: you need to add `artifacts/packages/$(Configuration)/Shipping` as a nuget source so the locally built packages can be picked up.

### Using it from VS

- No difference when running the EndToEnd tests in `in-tree` mode
- For switching to `out-of-tree` add `<TestsRunningOutsideRepo>true</TestsRunningOutsideRepo>` to `tests/Aspire.EndToEnd.Tests/Directory.props` *before* any imports.
    - tests cannot be run at this point as they will fail complaining about `artifacts/bin/dotnet-latest` being missing
    - Install the sdk+workload following the steps above
    - Run/debug the tests normally now, and they will be using the sdk
    - Also note that in this case the testproject is run from the bindir for `Aspire.EndToEnd.Tests`, so a path like `artifacts/bin/Aspire.EndToEnd.Tests/Debug/net8.0/testassets/testproject/`

### Using it from command line

- No difference when running the EndToEnd tests in `in-tree` mode
- Install the sdk+workload following the steps above
- When running the tests you can either:
    - set `<TestsRunningOutsideRepo>true</TestsRunningOutsideRepo>` to `tests/Aspire.EndToEnd.Tests/Directory.props` before any imports
    - or set the environment variable `TestsRunningOutsideRepo=true`

### Other details

- `tests/Shared/WorkloadTesting` - has sources essentially from `dotnet/runtime`'s `Wasm.Build.Tests`
- TestProject supports skipping individual resources like `--skip-resources oracledatabase,cosmos`
- `.runsettings` with 15min timeout
- `xunit.runner.json` - to get diagnostic messages
- The core infrastructure is not specific for `Aspire.EndToEnd.Tests`, and will get used by other test projects in the future

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2534)

### CODE REVIEW

**Comment by radical** [comment]
- [ ] Questions:
    - [ ] should a forced container cleanup be done at the end?
    - [ ] nuget-cache - Is it OK to expect the dev to manually delete aspire packages from the cache?
        - PR uses a new nuget-cache path for every run, which is slow on slow connections, but good on fast ones
    - [ ] Microsoft.NET.Sdk.Aspire: set `PackageType=''` to the manifest nuget is not`dotnetplatform`, and can be installed directly by the workload testing tasks. Check that nothing else is affected, and this is an acceptable change.

TODO:
    - [ ] nuget.helix.config - Use the root `nuget.config` and insert the built-nugets source

**Comment by joperezr** [comment]
Talked offline with @radical, marking as ready for review as this is ready for feedback.

File: tests/Shared/WorkloadTesting/ProcessExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+
+namespace Aspire.Workload.Tests;
+
+internal static class ProcessExtensions
+{
+    // private static int RunProcessAndWaitForExit(string fileName, string arguments, TimeSpan timeout, out string stdout)
```
Comment (eerhardt) [inline tests/Shared/WorkloadTesting/ProcessExtensions.cs:10]: This should be deleted.

File: tests/Shared/WorkloadTesting/BuildEnvironment.cs [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+        if (!string.IsNullOrEmpty(EnvironmentVariables.SdkForWorkloadTestingPath))
+        {
+            // always allow overridding the dotnet used for testing
+            sdkForWorkloadPath = EnvironmentVariables.SdkForWorkloadTestingPath;
+        }
+
+        HasSdkWithWorkload = expectSdkWithWorkload;
+        sdkForWorkloadPath = Path.GetFullPath(sdkForWorkloadPath);
+        DefaultBuildArgs = string.Empty;
```
Comment (eerhardt) [inline tests/Shared/WorkloadTesting/BuildEnvironment.cs:116]: Does this ever get set to anything other than `string.Empty`?

File: tests/Aspire.EndToEnd.Tests/IntegrationServicesTests.cs [hunk 3]
```diff
@@ -1,94 +1,111 @@
...
-            return null!;
+            // We wait until timeout for the /health endpoint to return successfully. We assume
+            // that components wired up into this project have health checks enabled.
+            await _integrationServicesFixture.IntegrationServiceA.WaitForHealthyStatusAsync("http");
+        }
+        catch
+        {
+            _testOutput.WriteLine($"[{DateTime.Now}] <<<< FAILED VerifyHealthyOnIntegrationServiceA --");
+            await _integrationServicesFixture.DumpDockerInfoAsync();
+            throw;
```
Comment (eerhardt) [inline tests/Aspire.EndToEnd.Tests/IntegrationServicesTests.cs:108]: Is it possible to do this in a Dispose() method on the class? That way it isn't duplicated across all the tests in the class.

File: tests/Aspire.EndToEnd.Tests/Aspire.EndToEnd.Tests.csproj [hunk 4]
```diff
@@ -2,10 +2,62 @@
 
   <PropertyGroup>
     <TargetFramework>$(NetCurrent)</TargetFramework>
+
+    <!-- no docker support on helix/windows yet -->
+    <RunTestsOnHelix Condition="'$(OS)' != 'Windows_NT'">true</RunTestsOnHelix>
+    <SkipTests Condition="'$(OS)' == 'Windows_NT'">true</SkipTests>
```
Comment (eerhardt) [inline tests/Aspire.EndToEnd.Tests/Aspire.EndToEnd.Tests.csproj:8]: Can we just say `RunTestsOnHelix=true` all the time, and then do these checks inside the tests? Similar to how I did it in RabbitMQ in https://github.com/dotnet/aspire/pull/2345. It just feels cleaner that way.

File: tests/Aspire.EndToEnd.Tests/Aspire.EndToEnd.Tests.csproj [hunk 5]
```diff
@@ -2,10 +2,62 @@
...
+
+    <TestUsingWorkloads>true</TestUsingWorkloads>
+    <InstallWorkloadForTesting>true</InstallWorkloadForTesting>
+
+    <BuiltNuGetsDir>$(ArtifactsShippingPackagesDir)</BuiltNuGetsDir>
+    <PackageVersionForWorkloadManifests>$(PackageVersion)</PackageVersionForWorkloadManifests>
+    <DefineConstants Condition="'$(TestsRunningOutOfTree)' == 'true'">TESTS_RUNNING_OUT_OF_TREE;$(DefineConstants)</DefineConstants>
```
Comment (eerhardt) [inline tests/Aspire.EndToEnd.Tests/Aspire.EndToEnd.Tests.csproj:15]: Can we get this from an env var when the test runs so we don't have to have an `#if` in the code?

File: eng/pipelines/templates/BuildAndTest.yml [hunk 6]
```diff
@@ -42,10 +43,22 @@ steps:
...
+    - ${{ if ne(parameters.isWindows, 'true') }}:
+      # E2E tests are disabled on windows for now
+      - script: ${{ parameters.dotnetScript }}
+                build
+                -restore
+                -p:ContinuousIntegrationBuild=true
+                -p:Configuration=${{ parameters.buildConfig }}
+                -t:InstallWorkloadUsingArtifacts
+                -bl:${{ parameters.repoLogPath }}/WorkloadInstallForTesting.binlog
+                $(Build.SourcesDirectory)/tests/Aspire.EndToEnd.Tests/Aspire.EndToEnd.Tests.csproj
```
Comment (eerhardt) [inline eng/pipelines/templates/BuildAndTest.yml:55]: Is it weird to pick the EndToEnd tests to have this "InstallWorkloadUsingArtifacts" target? It seems like something that should be coming from an "infrastructure" project. Not one of the test projects.

File: tests/testproject/TestProject.AppHost/TestProgram.cs [hunk 7]
```diff
@@ -5,12 +5,29 @@
...
 using System.Text.Json.Nodes;
 using Aspire.Hosting.Lifecycle;
+using Aspire.TestProject;
 
 public class TestProgram : IDisposable
 {
     private TestProgram(string[] args, Assembly assembly, bool includeIntegrationServices, bool includeNodeApp, bool disableDashboard)
     {
-        if (args.Contains("--disable-dashboard"))
+        IList<TestResourceNames> resourcesToSkip = [];
```
Comment (eerhardt) [inline tests/testproject/TestProject.AppHost/TestProgram.cs:14]: Do we actually need this `resourcesToSkip`? If we just skipped the cosmos and oracle tests, wouldn't that be enough? Yes it would still spin up those containers during the tests, but nothing would try to connect to them.

I don't really like all these if conditions everywhere. It really clutters up the code.

File: tests/Aspire.EndToEnd.Tests/IntegrationServicesFixture.cs [hunk 8]
```diff
@@ -2,40 +2,92 @@
...
 
+    private Process? _appHostProcess;
+    private readonly TaskCompletionSource _appExited = new();
+    private Dictionary<string, ProjectInfo>? _projects;
+    private readonly IMessageSink _diagnosticMessageSink;
+    private readonly TestOutputWrapper _testOutput;
+
+    public IntegrationServicesFixture(IMessageSink messageSink)
+    {
+        _diagnosticMessageSink = messageSink;
```
Comment (eerhardt) [inline tests/Aspire.EndToEnd.Tests/IntegrationServicesFixture.cs:42]: (nit) ctor parameter names should match the field/property names.

File: tests/Aspire.EndToEnd.Tests/IntegrationServicesTests.cs (hunk 3 above)
Comment (radical) [reply tests/Aspire.EndToEnd.Tests/IntegrationServicesTests.cs:108]: Yes, but then it will run for passing tests too. If that's acceptable then I'll move it there.

File: eng/pipelines/templates/BuildAndTest.yml (hunk 6 above)
Comment (radical) [reply eng/pipelines/templates/BuildAndTest.yml:55]: Yes, I was thinking of fixing it in a follow up though. But I can try to do it here too.

File: tests/testproject/TestProject.AppHost/TestProgram.cs (hunk 7 above)
Comment (radical) [reply tests/testproject/TestProject.AppHost/TestProgram.cs:14]: We have one test that health checks on the whole thing, and that would continue to fail.

> I don't really like all these if conditions everywhere. It really clutters up the code.
agree!

File: tests/Aspire.EndToEnd.Tests/Aspire.EndToEnd.Tests.csproj (hunk 5 above)
Comment (radical) [reply tests/Aspire.EndToEnd.Tests/Aspire.EndToEnd.Tests.csproj:15]: I had to add this just for the VS run, so you need to set only one property in one file to have the tests switch.
Using env var works perfectly well when running from command line, but I don't know what the workflow can be for setting it in case of VS. Open to suggestions.

File: tests/Aspire.EndToEnd.Tests/Aspire.EndToEnd.Tests.csproj (hunk 4 above)
Comment (radical) [reply tests/Aspire.EndToEnd.Tests/Aspire.EndToEnd.Tests.csproj:8]: Yes we can do it like that also. The reason for "also" is that if we don't have any tests that would get run, then not skipping the whole thing would mean:
1. installing sdk+workload - extra step on CI
2. sending this work item to helix with the big payload (sdk) for nothing to run, which dnceng kinda frowns upon.

I think we can add the attributes also, and for the specific case (windows) where nothing can run, we skip in the project file too. Thoughts?

File: tests/Shared/WorkloadTesting/BuildEnvironment.cs (hunk 2 above)
Comment (radical) [reply tests/Shared/WorkloadTesting/BuildEnvironment.cs:116]: No, I can remove this. Though this would come in handy when adding build tests which will be new'ing up projects. Not needed right now though.

File: NuGet.config [hunk 9]
```diff
@@ -12,13 +12,22 @@
...
     <add key="dotnet-public" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet-public/nuget/v3/index.json" />
     <add key="dotnet-eng" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet-eng/nuget/v3/index.json" />
     <add key="dotnet8" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet8/nuget/v3/index.json" />
     <add key="dotnet-libraries" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet-libraries/nuget/v3/index.json" />
     <add key="orleans-nightly" value="https://orleans.pkgs.visualstudio.com/orleans-public/_packaging/orleans-nightly/nuget/v3/index.json" />
+    <add key="dotnet9-transport" value="https://dnceng.pkgs.visualstudio.com/public/_packaging/dotnet9-transport/nuget/v3/index.json" />
   </packageSources>
   <packageSourceMapping>
+    <!-- used for installing test workloads for resolving aspire workload manifest -->
+    <packageSource key="nuget-local">
```
[the remark this answers is not in the document]
Comment (radical) [reply NuGet.config:25]: https://github.com/dotnet/runtime/blob/f135699310bccfa83b3b441d2f4f1571fe8df480/src/tasks/WorkloadBuildTasks/InstallWorkloadFromArtifacts.cs#L262

I agree that it does seem odd that you can't directly see what this links up to in the repo. I can have the package source name be passed to the task, which would be in aspire repo.

File: tests/Aspire.EndToEnd.Tests/IntegrationServicesTests.cs (hunk 3 above)
Comment (eerhardt) [reply tests/Aspire.EndToEnd.Tests/IntegrationServicesTests.cs:108]: Maybe just refactor the code so it isn't duplicated in all the tests if we can't tell if the test failed or not in Dispose.

File: tests/Aspire.EndToEnd.Tests/Aspire.EndToEnd.Tests.csproj (hunk 4 above)
Comment (eerhardt) [reply tests/Aspire.EndToEnd.Tests/Aspire.EndToEnd.Tests.csproj:8]: How long until we are able to run these tests on Windows?

File: tests/Shared/WorkloadTesting/BuildEnvironment.cs (hunk 2 above)
Comment (radical) [reply tests/Shared/WorkloadTesting/BuildEnvironment.cs:116]: I tried removing it, and it triggers some other refactoring. Can we keep this in though, as it will be useful when we add more tests?

File: eng/Version.Details.xml [hunk 10]
```diff
@@ -115,6 +115,10 @@
       <Sha>b4fa7f2e1e65ef49881be2ab2df27624280a8c55</Sha>
       <SourceBuild RepoName="source-build-reference-packages" ManagedOnly="true" />
     </Dependency>
+    <Dependency Name="Microsoft.NET.Runtime.WorkloadTesting.Internal" Version="9.0.0-preview.3.24151.4">
```
Comment (joperezr) [inline eng/Version.Details.xml:118]: NIT: Can we move this new dependency up to the section where we have the rest of the dotnet-runtime repo dependencies?

File: NuGet.config (hunk 9 above)
[the remark this answers is not in the document]
Comment (joperezr) [reply NuGet.config:26]: Does this throw a warning at all on the initial restore before the task has a chance to replace the source? Did we also consider manually passing in `-s <path>` to a manual invocation of restore instead of having this in the config?

File: eng/Versions.props [hunk 11]
```diff
@@ -7,6 +7,7 @@
     <VersionPrefix>8.0.0</VersionPrefix>
     <PreReleaseVersionLabel>preview.5</PreReleaseVersionLabel>
     <UseVSTestRunner>true</UseVSTestRunner>
+    <MicrosoftNETRuntimeWorkloadTestingInternalVersion>9.0.0-preview.3.24151.4</MicrosoftNETRuntimeWorkloadTestingInternalVersion>
```
Comment (joperezr) [inline eng/Versions.props:10]: NIT: please move bellow with the rest of the package versions.

File: eng/workloads/workloads.csproj [hunk 12]
```diff
@@ -10,8 +10,6 @@
     <VSTemp>$(WorkloadIntermediateOutputPath)VS/</VSTemp>
     <WorkloadOutputPath>$(ArtifactsBinDir)workloads/</WorkloadOutputPath>
     <WorkloadOutputPath Condition="'$(workloadArtifactsPath)' != ''">$(workloadArtifactsPath)/</WorkloadOutputPath>
-    <PackageSource>$(ArtifactsShippingPackagesDir)</PackageSource>
-    <PackageSource Condition="'$(workloadPackagesPath)' != ''">$(workloadPackagesPath)/</PackageSource>
```
Comment (joperezr) [inline eng/workloads/workloads.csproj:14]: Does this mean that this cannot get overriden any longer? cc: @danegsta in case he is aware if we depend on this.

Comment (radical) [reply eng/workloads/workloads.csproj:14]: Where does `$(workloadPackagesPath)` get set? I didn't find any reference to it in the repo.

File: eng/dcppack/dcppack.csproj [hunk 13]
```diff
@@ -39,8 +39,16 @@
 
   <Target Name="Build" />
 
+  <Target Name="AddPackageFiles" BeforeTargets="_GetPackageFiles">
+    <ItemGroup>
+      <DcpFiles Include="$(NuGetPackageRoot)microsoft.developercontrolplane.$(DcpPlatform)/$(MicrosoftDeveloperControlPlanedarwinamd64PackageVersion)/tools/**/*" />
```
Comment (joperezr) [inline eng/dcppack/dcppack.csproj:44]: NIT: Can we use the GeneratePackagePath property instead? Or does that not work when using PackageDownload?

File: eng/dashboardpack/dashboardpack.csproj [hunk 14]
```diff
@@ -36,7 +38,7 @@
     </ItemGroup>
 
     <MakeDir Directories="$(DashboardPublishedArtifactsOutputDir)/$(DashboardRuntime)" />
-    <ZipDirectory SourceDirectory="$(DotNetOutputPath)Aspire.Dashboard/$(Configuration)/$(TargetFramework)/$(DashboardRuntime)/publish" DestinationFile="$(DashboardPublishedArtifactsOutputDir)/$(DashboardRuntime)/aspire-dashboard-$(DashboardRuntime).zip" />
+    <ZipDirectory SourceDirectory="$(DotNetOutputPath)Aspire.Dashboard/$(Configuration)/$(TargetFramework)/$(DashboardRuntime)/publish" DestinationFile="$(DashboardPublishedArtifactsOutputDir)/$(DashboardRuntime)/aspire-dashboard-$(DashboardRuntime).zip" Overwrite="true" />
```
Comment (joperezr) [inline eng/dashboardpack/dashboardpack.csproj:41]: Why do we need to override?

File: eng/dashboardpack/dashboardpack.csproj [hunk 15]
```diff
@@ -26,8 +26,10 @@
 
   <Target Name="Build" />
 
-  <Target Name="BeforeBuild" BeforeTargets="Build">
-    <MSBuild Projects="../../src/Aspire.Dashboard/Aspire.Dashboard.csproj" Targets="publish" Properties="Configuration=$(Configuration);Platform=$(Platform);TargetFramework=$(TargetFramework);RuntimeIdentifier=$(DashboardRuntime)" />
+  <Target Name="BeforeBuild" BeforeTargets="_GetPackageFiles">
+    <!-- separate restore step to allow targets from nugets to get imported in the publish step -->
+    <MSBuild Projects="../../src/Aspire.Dashboard/Aspire.Dashboard.csproj" Targets="restore" Properties="Configuration=$(Configuration);Platform=$(Platform);TargetFramework=$(TargetFramework);RuntimeIdentifier=$(DashboardRuntime)" />
+    <MSBuild Projects="../../src/Aspire.Dashboard/Aspire.Dashboard.csproj" Targets="publish" Properties="Configuration=$(Configuration);Platform=$(Platform);TargetFramework=$(TargetFramework);RuntimeIdentifier=$(DashboardRuntime);_Some_unique_prop=$([System.Guid]::NewGuid())" />
```
Comment (joperezr) [inline eng/dashboardpack/dashboardpack.csproj:32]: Can you explain why are we passing in this extra property? This is breaking incremental build. Essentially Restore is happening again given there are new global properties.

File: src/Microsoft.NET.Sdk.Aspire/Microsoft.NET.Sdk.Aspire.csproj [hunk 16]
```diff
@@ -5,6 +5,9 @@
   <PropertyGroup>
     <PackageId>$(PackageId).Manifest-$(DotNetAspireManifestVersionBand)</PackageId>
     <Description>.NET Aspire workload manifest</Description>
+
+    <!-- this allows installing the manifest as a nuget, which is useful for testing -->
+    <PackageType />
```
Comment (joperezr) [inline src/Microsoft.NET.Sdk.Aspire/Microsoft.NET.Sdk.Aspire.csproj:10]: Just Curious, does this change anything in the nuspec?

File: src/Shared/Workload.targets [hunk 17]
```diff
@@ -12,13 +14,7 @@
     <DotNetOutputPath>$(RepoRoot)artifacts/bin/</DotNetOutputPath>
     <DotNetDirectory>$(DotNetOutputPath)dotnet/</DotNetDirectory>
     <DotNetPacksDirectory>$(DotNetDirectory)packs/</DotNetPacksDirectory>
-    <VersionBand Condition=" '$(VersionBand)' == '' ">$([System.Text.RegularExpressions.Regex]::Match($(MicrosoftDotnetSdkInternalPackageVersion), `^\d+\.\d+\.\d`))00</VersionBand>
```
Comment (joperezr) [inline src/Shared/Workload.targets:15]: Just curious, why extract only those properties and not the rest?

Comment (radical) [reply src/Shared/Workload.targets:15]: The extracted properties define the version, and band used for the workload. The remaining properties are relevant to the various projects that import it for building packages. Some of these common properties could still be moved to the props file, but not all.

File: eng/workloads/workloads.csproj (hunk 12 above)
Comment (danegsta) [reply eng/workloads/workloads.csproj:14]: Should be fine as long as the workload still builds correctly

File: eng/dashboardpack/dashboardpack.csproj (hunk 15 above)
Comment (danegsta) [reply eng/dashboardpack/dashboardpack.csproj:32]: Is this solving a particular issue for the helix test scenario? Aspire.Dashboard should already be restored as part of the main build before this project runs.

File: eng/dashboardpack/dashboardpack.csproj (hunk 14 above)
Comment (radical) [reply eng/dashboardpack/dashboardpack.csproj:41]: We don't. I had added it for something else and forgot to remove!

File: eng/dashboardpack/dashboardpack.csproj (hunk 15 above)
Comment (radical) [reply eng/dashboardpack/dashboardpack.csproj]: This was a stale change which I had to add earlier because a restored nuget package was adding targets, but the project was not re-evaluated thus the targets were not available.
Subsequently, I had changed the way we install the workload for testing, but missed removing this change. Thank you for questioning it!

Comment (radical) [reply eng/dashboardpack/dashboardpack.csproj]: This wasn't stale, instead the conditions were reproducible only on CI. 
TL;dr - this was needed because it needed fresh restore, and a project re-evaluation to load targets from `grpc.tools` nuget.

Details:

On CI we used `./build.sh` for the main build, and then `.dotnet.sh build` to install the workload, which hit the target above.
- `build.sh` - sets `NUGET_PACKAGES` to `$repoRoot/.packages`
- and `./dotnet.sh build` defaulted to `$HOME/.nuget/packages`

So,
1. `./build.sh` - restores nugets to `$repoRoot/.packages`, and imports targets from the grpc.tools nuget.
2. `.dotnet build` - runs the above line and runs `publish` target
    a. which imports the `nuget.g.{props,targets}` for the E2E project, but it fails to import the `grpc.tools` targets because it is now looking at the wrong path to the restored package!
    b. Even if the `restore` target is run before this `publish`, the msbuild project is not re-evaluated and thus the `gprc.tools` targets don't get imported
Instead, if we `restore`, and then `publish` but with a new property added then it triggers a re-evaluation of the project, which causes the `grpc.tools` targets to get imported, and run

Now that I have figured out the reason for all this mess - `./build.sh vs ./dotnet.sh build`, I have:
1. changed to use `./build.sh` for both the cases
2. And have the target instead be triggered based on a property passed to the command on CI only.

This fixes the issue correctly.

File: src/Microsoft.NET.Sdk.Aspire/Microsoft.NET.Sdk.Aspire.csproj (hunk 16 above)
Comment (radical) [reply src/Microsoft.NET.Sdk.Aspire/Microsoft.NET.Sdk.Aspire.csproj:10]: Yes, this removes:
```xml
    <packageTypes>
      <packageType name="DotnetPlatform" />
    </packageTypes>
```
.. from the `nuspec`. I checked other manifest nugets and they do have this specified, but I skipped this for the mono/wasm manifests, and it hasn't been an issue.

In a follow up for `dotnet/runtime`, I can change the way the nuget is installed which would remove the need for `PackageType` to be `''`.

File: eng/dashboardpack/dashboardpack.csproj (hunk 14 above)
Comment (radical) [reply eng/dashboardpack/dashboardpack.csproj:39]: Scratch that. This is useful for local builds when you might build the project even when the zip already exists.

File: eng/dcppack/dcppack.csproj (hunk 13 above)
Comment (radical) [reply eng/dcppack/dcppack.csproj:44]: It doesn't seem to work for `PackageDownload`.

File: NuGet.config (hunk 9 above)
[the remark this answers is not in the document]
Comment (radical) [reply NuGet.config:26]: No warnings. Wouldn't `-s path` only add a package source but not the mapping, and because of that restoring from the source would fail. That is what I hit when I removed this mapping.

File: tests/nuget.helix.config [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+    <packageSource key="dotnet-libraries">
+      <package pattern="Microsoft.DeveloperControlPlane*" />
+      <package pattern="System.CommandLine" />
+    </packageSource>
+    <packageSource key="dotnet-eng">
+      <package pattern="*" />
+    </packageSource>
+    <packageSource key="orleans-nightly">
+      <package pattern="Microsoft.Orleans.*" />
+    </packageSource>
```
[the remark this answers is not in the document]
Comment (radical) [reply tests/nuget.helix.config]: `nuget.helix.config` is generated from the root `nuget.config` now.

File: NuGet.config (hunk 9 above)
[the remark this answers is not in the document]
Comment (radical) [reply NuGet.config:26]: Once https://github.com/dotnet/runtime/pull/99277 is merged this entry can be removed, and it would be added at build time.

File: tests/Aspire.EndToEnd.Tests/IntegrationServicesFixture.cs [hunk 19]
```diff
@@ -2,40 +2,92 @@
...
+            _testOutput.WriteLine("");
+        }
+
+        {
+            using var cmd = new DotNetCommand(BuildEnvironment, _testOutput, label: "build")
+                .WithWorkingDirectory(appHostDirectory);
+
+            (await cmd.ExecuteAsync(CancellationToken.None, $"build -bl:{Path.Combine(BuildEnvironment.LogRootPath, "testproject-build.binlog")} -v m"))
+                .EnsureSuccessful();
+        }
```
Comment (eerhardt) [inline tests/Aspire.EndToEnd.Tests/IntegrationServicesFixture.cs:76]: This is super confusing formatting. When you skim the code it looks like the `using var cmd =....` code only runs in the `else` of the `if (TestsRunningOutsideOfRepo)` condition.

Instead of this custom scoping, we should just extract lines 70-76 to a method. It will make this code much easier to read.

File: tests/Aspire.EndToEnd.Tests/IntegrationServicesFixture.cs [hunk 20]
```diff
@@ -2,40 +2,92 @@
...
         var output = new StringBuilder();
-        var appExited = new TaskCompletionSource();
         var projectsParsed = new TaskCompletionSource();
         var appRunning = new TaskCompletionSource();
         var stdoutComplete = new TaskCompletionSource();
         var stderrComplete = new TaskCompletionSource();
         _appHostProcess = new Process();
-        _appHostProcess.StartInfo = new ProcessStartInfo("dotnet", "run -- --disable-dashboard")
+
+        string processArguments = $"run --no-build -- ";
```
Comment (eerhardt) [inline tests/Aspire.EndToEnd.Tests/IntegrationServicesFixture.cs:85]: Why did we split build from run? Was it so we could split the binlog?

File: tests/Aspire.EndToEnd.Tests/IntegrationServicesFixture.cs [hunk 21]
```diff
@@ -2,40 +2,92 @@
...
+            _testOutput.WriteLine($"   Using dotnet: {BuildEnvironment.DotNet}");
+            _testOutput.WriteLine($"****************************************");
+            _testOutput.WriteLine("");
+        }
+
+        {
+            using var cmd = new DotNetCommand(BuildEnvironment, _testOutput, label: "build")
+                .WithWorkingDirectory(appHostDirectory);
+
+            (await cmd.ExecuteAsync(CancellationToken.None, $"build -bl:{Path.Combine(BuildEnvironment.LogRootPath, "testproject-build.binlog")} -v m"))
```
Comment (eerhardt) [inline tests/Aspire.EndToEnd.Tests/IntegrationServicesFixture.cs:74]: Why `-v m`?

File: tests/Aspire.EndToEnd.Tests/IntegrationServicesFixture.cs [hunk 22]
```diff
@@ -151,15 +287,30 @@ public async Task DisposeAsync()
...
+        if (_appHostProcess is null || _appHostProcess.HasExited || _appExited.Task.IsCompleted)
+        {
+            throw new InvalidOperationException("The app host process is not running.");
+        }
+    }
 
-        while (directory != null && !Directory.Exists(Path.Combine(directory, ".git")))
+    private static IList<string> GetResourcesToSkip()
+    {
+        List<string> resourcesToSkip = new();
```
Comment (eerhardt) [inline tests/Aspire.EndToEnd.Tests/IntegrationServicesFixture.cs:300]:
```suggestion
        HashSet<string> resourcesToSkip = new();
```

? We don't really want dupes in this collection.

File: tests/Aspire.EndToEnd.Tests/IntegrationServicesFixture.cs (hunk 21 above)
Comment (radical) [reply tests/Aspire.EndToEnd.Tests/IntegrationServicesFixture.cs:74]: Because most of the time the build is going to pass and doesn't give any useful info for issues in tests. If the build fails then those messages will show up, and a `testproject-build.binlog` will be available for debugging offline.

File: tests/Aspire.EndToEnd.Tests/IntegrationServicesFixture.cs (hunk 20 above)
Comment (radical) [reply tests/Aspire.EndToEnd.Tests/IntegrationServicesFixture.cs:85]: Yes

**Comment by radical** [comment]
@eerhardt I tried to add oracledb back, but it failed because the test timeout was 5minutes. I think we continue with that in a follow up PR, so reverting that change.

File: eng/dashboardpack/dashboardpack.csproj [hunk 23]
```diff
@@ -26,7 +26,7 @@
 
   <Target Name="Build" />
 
-  <Target Name="BeforeBuild" BeforeTargets="Build">
+  <Target Name="BeforeBuild" BeforeTargets="_GetPackageFiles">
```
Comment (joperezr) [inline eng/dashboardpack/dashboardpack.csproj:29]: Why do we need this change? In general, it's best to avoid depending on "private" tragets (Targets that start with an underscore)

Comment (radical) [reply eng/dashboardpack/dashboardpack.csproj:29]: I wanted to limit this target to only when we are generating a nuget.

Is this target needed only for generating the zip file? Is the zip file used for the dashboard docker image?

File: eng/dcppack/dcppack.csproj [hunk 24]
```diff
@@ -39,8 +39,16 @@
 
   <Target Name="Build" />
 
+  <Target Name="AddPackageFiles" BeforeTargets="_GetPackageFiles">
```
Comment (joperezr) [inline eng/dcppack/dcppack.csproj:42]: Same comment here, if we could instead hook up to one of the "public" targets: https://learn.microsoft.com/en-us/nuget/reference/msbuild-targets#advanced-extension-points-to-create-customized-package

File: src/Microsoft.NET.Sdk.Aspire/Microsoft.NET.Sdk.Aspire.csproj (hunk 16 above)
Comment (joperezr) [reply src/Microsoft.NET.Sdk.Aspire/Microsoft.NET.Sdk.Aspire.csproj:10]: mmm, I suppose this property is set via Arcade. It would be good to check what the implications of that are with someone from coreeng team.

**Review by joperezr — APPROVED** [review]
Added a few comments/suggestions but the rest LGTM.

File: eng/dashboardpack/dashboardpack.csproj (hunk 23 above)
Comment (joperezr) [reply eng/dashboardpack/dashboardpack.csproj:29]: This target mainly accomplishes two things:

- Adds the right content to the each of the dashboard packs (which actually may no longer be needed now that we have the dashboard as a tool) cc: @danegsta 
- Zips up the published output as that will be uploaded as a blob so that it can then be used to build the docker image for the dashboard.

Comment (radical) [reply eng/dashboardpack/dashboardpack.csproj]: Reverted.

File: eng/pipelines/templates/BuildAndTest.yml [hunk 25]
```diff
@@ -21,6 +21,7 @@ parameters:
 steps:
   - script: ${{ parameters.buildScript }}
             -restore -build
+            -pack
```
[the remark this answers is not in the document]
Comment (radical) [reply eng/pipelines/templates/BuildAndTest.yml:24]: Good point. This pack could conflict with the subsequent pack, since that one is signed. I will investigate this in a separate PR. And it will really be needed only when we have this running on windows again.
For now, I will add `-pack` only for linux, where the separate `pack` step is not defined at all.

**Comment by radical** [comment]
Follow-up work is being tracked in https://github.com/dotnet/aspire/issues/2687 .
